### PR TITLE
load map.php from theme/translations/map.php

### PR DESCRIPTION
**Pull request by alfredbez**
allows us to create a single map.php for all languages

### CODE REVIEW

**Comment by alfredbez** [comment]
Currently we need a map.php for every language. This may look like this:
```
de
├── cust_lang.php
├── lang.php
├── map.php
└── theme_options.php
en
├── cust_lang.php
├── lang.php
├── map.php
└── theme_options.php
fr
├── cust_lang.php
├── lang.php
├── map.php
└── theme_options.php
```
Afterwards we can use only one map.php (which is located in the new `lang` folder) for all languages:
```
de
├── cust_lang.php
├── lang.php
└── theme_options.php
en
├── cust_lang.php
├── lang.php
└── theme_options.php
fr
├── cust_lang.php
├── lang.php
└── theme_options.php
lang
└── map.php
```
We could even move all language/translations-files to that folder someday:
```
lang
├── map.php
├── de
│   ├── cust_lang.php
│   ├── lang.php
│   └── theme_options.php
├── en
│   ├── cust_lang.php
│   ├── lang.php
│   └── theme_options.php
└── fr
    ├── cust_lang.php
    ├── lang.php
    └── theme_options.php
```
It is also possible to overwrite the `map.php` for specific languages, the last existing and readable file from the `$mapFiles` will be used as the mapping file.

```php
$mapFiles = [
  $config->getAppDir() . '/views/' . $theme . '/lang/map.php',
  $config->getAppDir() . '/views/' . $theme . '/' . $languageAbbr . '/map.php',
  $config->getAppDir() . '/views/' . $customTheme . '/lang/map.php',
  $config->getAppDir() . '/views/' . $customTheme . '/' . $languageAbbr . '/map.php',
];

foreach ($mapFiles as $tmpMapFile) {
  if (file_exists($tmpMapFile) && is_readable($tmpMapFile)) {
    $mapFile = $tmpMapFile;
  }
}
```

Should I rename the `lang` directory to `translations`?

**Comment by keywan-ghadami-oxid** [comment]
I like this pull request, code is well writen and it is backward compatible.
But about your question to rename the folder i am not sure, what would be the benefit?

**Comment by alfredbez** [comment]
>But about your question to rename the folder i am not sure, what would be the benefit?

There is currently no `lang` and no `translations` folder in a theme. This PR adds a new folder and I named it `lang`, but we can choose a different name like `translations`.

**Comment by keywan-ghadami-oxid** [comment]
Ok now i got it :-). 
Having only the map.php in that new folder is something i do not like much.
You could simply move map.php one level up without having a new folder - but of cause it is somehow unstructured so i vote for a 'translations' folder having all translation files in it because then it is similar to the translation folder in the 'Application' folder. And i think 'translations' is more expressive then 'lang'.

**Comment by alfredbez** [comment]
I changed `lang` to `translations` now.

I'd suggest the following folder-structure for themes (just like for modules):
```
translations
├── map.php
├── de
│   ├── cust_lang.php
│   ├── lang.php
│   └── theme_options.php
├── en
│   ├── cust_lang.php
│   ├── lang.php
│   └── theme_options.php
└── fr
    ├── cust_lang.php
    ├── lang.php
    └── theme_options.php
```

**Comment by robertblank** [comment]
I like the idea of having only one map file. 
Is there a use case at all for having one map per language?
If not I would not permit to make it overridable per language.

I also like the idea of having a translations folder inside the theme.
I will promote the idea here in the team.

**Comment by alfredbez** [comment]
This would be very useful, especially for multilanguage projects (we have nearly 30 languages in our project)

File: source/Core/Language.php [hunk 1]
```diff
@@ -991,37 +991,43 @@ protected function _getLanguageFileData($blAdmin = false, $iLang = 0, $aLangFile
...
+            $mapFiles = [
+                $config->getAppDir() . '/views/' . $theme . '/translations/map.php',
+                $config->getAppDir() . '/views/' . $theme . '/' . $languageAbbr . '/map.php',
+                $config->getAppDir() . '/views/' . $customTheme . '/translations/map.php',
+                $config->getAppDir() . '/views/' . $customTheme . '/' . $languageAbbr . '/map.php',
+            ];
+
+            foreach ($mapFiles as $tmpMapFile) {
+                if (file_exists($tmpMapFile) && is_readable($tmpMapFile)) {
+                    $mapFile = $tmpMapFile;
```
Comment (Sieg) [inline source/Core/Language.php]: Looks like we could reorder the $mapFiles and break the cycle on the first one we will find?

Comment (alfredbez) [reply source/Core/Language.php]: yes good idea

**Comment by alfredbez** [comment]
I've rebased it, please take a look at this. If you want I can also remove the possibility to have a `map.php` inside a language folder, as @robertblank [said](https://github.com/OXID-eSales/oxideshop_ce/pull/449#issuecomment-247643343).

This should already work:
```diff
--- a/source/Core/Language.php
+++ b/source/Core/Language.php
@@ -1015,9 +1015,7 @@ class Language extends \OxidEsales\Eshop\Core\Base
             $customTheme = $this->getRealThemeName($config->getConfigParam("sCustomTheme"), $isAdmin);
 
             $mapFiles = [
-                $this->getLanguageMappingFilePath($customTheme, $language),
                 $this->getLanguageMappingFilePath($customTheme),
-                $this->getLanguageMappingFilePath($theme, $language),
                 $this->getLanguageMappingFilePath($theme),
             ];
 
@@ -1043,15 +1041,14 @@ class Language extends \OxidEsales\Eshop\Core\Base
      *
      * @return string
      */
-    private function getLanguageMappingFilePath($themeName, $language = null)
+    private function getLanguageMappingFilePath($themeName)
     {
         $config = \OxidEsales\Eshop\Core\Registry::getConfig();
-        $mapFolder = $language === null ? 'translations' : Registry::getLang()->getLanguageAbbr($language);
 
         return $config->getAppDir() . DIRECTORY_SEPARATOR .
             'views' . DIRECTORY_SEPARATOR .
             $themeName . DIRECTORY_SEPARATOR .
-            $mapFolder . DIRECTORY_SEPARATOR .
+            'translations' . DIRECTORY_SEPARATOR .
             'map.php';
     }
```

**Comment by Sieg** [comment]
Hi @alfredbez, i have merged the possibility to have one map.php file to our master at 3944e1fc3de5cf9119f33cb5d27921a459a31d86 with several changes to your code from my side. Let me know if something is wrong.

**Pull request closed without merging**